### PR TITLE
Fixes #894 - Fix networking logic issues that caused fuel tank explosions to not fully register

### LINKED ISSUE #894: Clients cannot walk on tiles where explosions took place
# Preliminary information

## Category
BUG

## Bug: Release or Development?
0.2.2 beta 4 steam

# Report
## Current Behavior
You cannot step on tiles (you lerp back) explosives were located on (prior to their explosion)

## Expected/Wanted/Requested Behavior
Want to walk free

## Steps to Reproduce
Please enter the steps to reproduce the bug or behaviour:
1. Explode a fuel tank from safe distance
2. Try to step on the tile where the tank was

## OS used:
OSX 10.9

## System specifications
MBP mid2010

## Network
beta server 60ms

### PR DESCRIPTION
**Pull request by chairbender**
### Purpose
Fix networking logic issues that caused fuel tank explosions to not fully register. Add lots more comments as well.

Fuel tanks will not explode anymore due to the health refactor #1397 , but I still fixed the underlying issue in the RegisterTile / RegisterObject behavior which may fix other latent issues. A new issue will be created to cover adding an object health behavior.

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer

### Notes:
This was mostly caused by RegisterTile.SetParent() not handling the case where the current position is TransformState.HiddenPos. Also, there was a lot of custom logic in ExplodeWhenShot attempting to hide / disable collisions on the object when it should've just called DisappearFromWorld/DisappearFromWorldServer instead.

This issue showed up in headless but also in normal client / server. You can validate headless on your local machine pretty easily just by invoking the player from the command line like so. Works for me on Windows:
.\unitystation - 0.3.1 OutpostStation.exe -batchmode -
nographics -logfile log2.txt

I tried squashing the commits this time, I hope I did it properly (looks like I may have missed one).

### CODE REVIEW

**Comment by chairbender** [comment]
Marking DNM right now as I discovered the merge caused some unneeded changes to be checked in.

**Comment by chairbender** [comment]
Ready to PR now, please review.

**Comment by chairbender** [comment]
ah nice, cleaned up the non .cs files that were showing as modified in this PR. This PR should only need .cs file changes and nothing else. Still ready for review.

File: UnityProject/Assets/Scripts/Objects/ExplodeWhenShot.cs [hunk 1]
```diff
@@ -121,29 +146,27 @@ private bool NotSameObject(Collider2D localCollider)
...
+		//NOTE: This runs on both the client and the server.
+		//When something goes boom on a client, the client makes its own version go boom,
+		// the server makes its own version of the object go boom and
+		//sends an RPC to all the clients to tell them to make their version of the object go boom as well. So this
+		//method ends up being invoked on clients and server.
+
+		
+		//Shake if the player is on the same matrix (check for null in case this is a headless server)
+		if (PlayerManager.LocalPlayer != null &&
+			PlayerManager.LocalPlayer.gameObject.GetComponent<RegisterPlayer>() == registerObject.Matrix)
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/ExplodeWhenShot.cs:163]: I think later on we will need a distance limit for the size of the explosion. So a small explosions doesn't rattle people on the other side of the station

File: UnityProject/Assets/Scripts/Objects/ExplodeWhenShot.cs [hunk 2]
```diff
@@ -121,29 +146,27 @@ private bool NotSameObject(Collider2D localCollider)
...
-			Logger.LogWarning("Object may of already been removed", Category.Health);
-		}
-
-		foreach (Collider2D collider2d in gameObject.GetComponents<Collider2D>())
-		{
-			collider2d.enabled = false;
-		}
 
-		string name = explosions[Random.Range(0, explosions.Length)];
+		// Instantiate a clone of the source so that multiple explosions can play at the same time.
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Objects/ExplodeWhenShot.cs:168]: ahh very nice. I still need to work out a generic version of this to put on SoundManager

File: UnityProject/Assets/Scripts/PlayGroups/PlayerSync.cs [hunk 3]
```diff
@@ -6,441 +6,481 @@
...
-	public struct PlayerState
-	{
-		public bool Active => Position != TransformState.HiddenPos;
 
-		public int MoveNumber;
-		public Vector3 Position;
+/// <summary>
+/// Holds player state information used for interpolation, such as player position, flight direction etc.
+/// Gives client enough information to be able to smoothly interpolate the player position.
+/// </summary>
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/PlayGroups/PlayerSync.cs:13]: Oh wow, you documented PlayerSync. Thanks for that!

File: UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterObject.cs [hunk 4]
```diff
@@ -1,44 +1,46 @@
...
+		return Passable || Position == TransformState.HiddenPos;
+	}
 
-		public override bool IsAtmosPassable()
-		{
-			return AtmosPassable;
-		}
+	public override bool IsAtmosPassable()
+	{
+		return AtmosPassable || Position == TransformState.HiddenPos;
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Scripts/Tilemaps/Behaviours/Objects/RegisterObject.cs:23]: cool makes sense

File: UnityProject/Assets/Tests/ExplodeWhenShotTest.cs [hunk 5]
```diff
@@ -51,7 +51,7 @@ public IEnumerator Should_Destroy_Bullet()
 	[UnityTest]
```
Comment (DooblyNoobly) [inline UnityProject/Assets/Tests/ExplodeWhenShotTest.cs:51]: We haven't enforced Unit Testing for awhile (as it became a hassle to maintain with the rapid changes) but thanks for adjusting this also.

**Review by DooblyNoobly — APPROVED** [review]
Thanks for all the additional documentation! Lets discuss what is needed now for a new ObjectHealth component on discord then I'll create a bounty for it.

**Pull request merged**